### PR TITLE
Update "Your first 3D game" tutorial for Godot 4

### PR DESCRIPTION
**Pull request by TheYellowArchitect**
- I updated all non-4.0-related images (e.g. showed `Translation` instead of `Position`)
- Added some much-needed images
- Fixed all code to work for 4.0 (except 1 thing, see below)
- Polished many minor things which should have been polished even in the 3.x branch, like hyperlinking/referencing class names and formating node names (instead of _italic_)

Some images are in .webp format because I couldn't get .png or .jpg files in the size of the original images. Otherwise all new images would be 2~3 times bigger like https://github.com/godotengine/godot-docs/pull/6235
So, if docs don't support .webp, don't merge this pull request.

There are so many changes, that I think it's bloated to make a changelog for each (see changed files). 
Edit: Below is fixed/solved

-----------------

~~Instead, I will present the only problem which keeps this from being merged (so I guess it's a draft because of 3 codelines lol)~~

~~Scroll down into the big codeblock of this
https://docs.godotengine.org/en/latest/getting_started/first_3d_game/06.jump_and_squash.html#coding-the-squash-mechanic~~

```python
for index in range(get_slide_count()):
  # We check every collision that occurred this frame.
  var collision = get_slide_collision(index)
  # If we collide with a monster...
  if collision.collider.is_in_group("mob"):
      var mob = collision.collider
      # ...we check that we are hitting it from above.
      if Vector3.UP.dot(collision.normal) > 0.1:
          # If so, we squash it and bounce.
          mob.squash()
          velocity.y = bounce_impulse
```

~~I converted it to this:~~
```python
# Iterate through all collisions that occured this frame
# in C this would be for(int i = 0; i < collisions.Count; i++)
for index in range(get_slide_collision_count()):
    # We get one of the collisions with the player
    var collision = get_slide_collision(index)

    # If the collision is with ground
    if (collision.get_collider(index) == null):
	return

    # If the collider is with a mob
    if collision.get_collider(index).is_in_group("mob"):
	var mob = collision.get_collider(index)
	# we check that we are hitting it from above.
	if Vector3.UP.dot(collision.get_normal(index)) > 0.1:
		# If so, we squash it and bounce.
		mob.squash()
		target_velocity.y = bounce_impulse
```

~~Note that null conditional. Basically I cannot bugfix it, because I am a newbie and don't know how `move_and_slide()` really works. Somehow, it gets index+1 and it goes over the range and gives null. So if you run the above code, game runs normally but if you try to move, it gives error. If you jump and move, it doesn't give error (until you are back on the ground, so it's definitely something to do with ground collision)~~

--------------

The other thing is the animation tutorial, which just by changing transform/rotation as it said, makes the player and mobs who play these animations glitch (player cannot move or jump, mobs spawn at center)

If you want to debug it, I have uploaded the reproduction project:
[Squash40Demo.tar.gz](https://github.com/godotengine/godot-docs/files/9677294/Squash40Demo.tar.gz)

Basically, once those 2 problems are fixed, this can work as the final/ready project 4.0 for this demo.

To start debugging the first bug (ground collision), select the AnimationPlayer in Player and Mob scenes, and turn off the autoplay+loop. No idea about the animation bug (if ground collision is fixed, without animations it plays perfectly).

In case someone argues to not merge this because this project sounds broken, I want to note that this pull request, even in this state (1 animation bug) is far better than the master branch 3D Game Tutorial which is full of errors and misnames (over 10 errors)

### CODE REVIEW

**Comment by TheYellowArchitect** [comment]
I built this and read it from start to finish. I did the above minor polishes while I was doing that, fixing typos and whatnot. Anyway, I confirmed it really is ready to merge. Even with the 2 bugs mentioned above, the image additions alone elevate the quality of the tutorial. And I am willing to update the images when the styling is final (e.g. "Node3d" to "Node3D")

P.S. 5 images in total are very compressed (quality 80 wasn't kind to them), and I will re-upload cleaner versions when merged

**Comment by mhilbrunner** [comment]
Added discussion for now re: .webp, from a brief discussion in the #documentation channel on the official RC, most docs maintainers/contributors see the value of moving to .webp, even pretty soon-ish, but there are downsides in bloating the Git repo and also not all tooling supporting .webp yet.

In summary: move to .webp has consensus, _when_ and _how_ is mostly the question.

**Comment by Calinou** [comment]
`.webp` can be lossless, and lossless will be used when the file size isn't too large (less than 250 KB or so).

**Comment by TheYellowArchitect** [comment]
`.webp` can use lossless compression, but to have no quality degradation, the `.png` images which are to be converted into `.webp` must be uncompressed (at least from what I remember, cant confirm)

If some `.png` images are compressed and the above is true, then I suggest doing a backup of all `.png` images in some new repo, for the eventual `.avif` conversion

Slightly off-topic: The user making a doc page should be allowed to post an image in `.jpg`,`.png`,`.webp` and not forced to upload exclusively `.webp`, because there are still many who use those formats, e.g. when they take a snapshot of their computer their images are `.jpg`/`.png`
Indeed, everyone knows of `.webp` but forcing users to convert images (many don't know how to batch convert, or will use online image converters lol) will result in slightly more effort from the user and hence less images in docs from new users
So, even if the user posts `.jpg` or `.png`, the image should be left either as they are, or auto-converted to `.webp` on building the docs

**Comment by mhilbrunner** [comment]
>  because there are still many who use those formats, e.g. when they take a snapshot of their computer their images are .jpg/.png

Currently, we ask all PNGs to go through a PNG compression tool (if they're not very small). It's IMO not more of a ask to convert to .webp. instead.

**Comment by TheYellowArchitect** [comment]
>Currently, we ask all PNGs to go through a PNG compression tool (if they're not very small). It's IMO not more of a ask to convert to .webp. instead.

I understand, but there is no downside to allowing `.jpg` and `.png` alongside `.webp`, so why not help (new) contributors, by allowing them to upload their images? It's not something **required** to work like asking a contributor to rebase after a few merges.
The `.jpg` `.png` images could even be automatically converted to `.webp` and it's the best of both worlds

I think the best way to decide the above, would be to have some data of how many images are uploaded by old/veteran users. If it's 80+%, then forcing exclusively `.webp` is the best path, as all of those users are familiar with `.webp`
I assume the above is true, since all new doc pages are made by veteran users who made new godot features lol
Anyway, as long the original uncompressed `.png` images are backed up somewhere, and as long `.webp` is allowed, any choice regarding `.webp` is beneficial imo

File: getting_started/first_3d_game/09.adding_animations.rst [hunk 1]
```diff
@@ -242,10 +254,10 @@ Animating the mobs
...
 node structure, you can copy them to different scenes.
 
-For example, both the *Mob* and the *Player* scenes have a *Pivot* and a
-*Character* node, so we can reuse animations between them.
+For example, both the ``Mob`` and the ``Player`` scenes have a ``Pivot`` and a
+``Character`` node, so we can reuse animations between them.
 
-Open the *Player* scene, select the animation player node and open the "float" animation.
+Open the ``Player.tscn`` scene, select the ``AnimationPlayer`` node and open the "float" animation.
 Next, click on **Animation > Copy**. Then open ``Mob.tscn`` and open its animation
```
Comment (Sonix) [inline getting_started/first_3d_game/09.adding_animations.rst:261]: I wasn't able to find these buttons, however I was able to find similar commands in the Animation -> Manage Animations menu, though I'm not sure thats a good / recommended way to copy/paste animations.

**Review by Sonix — COMMENTED** [review]
I followed this with Godot 4.0 beta 3 and could mostly finish without problems, good work! I have a few notes / changes for which I'll provide a small PR shortly.

File: getting_started/first_3d_game/09.adding_animations.rst (hunk 1 above)
Comment (TheYellowArchitect) [reply getting_started/first_3d_game/09.adding_animations.rst:261]: You are right. I am trying to find these buttons myself, and it seems I meant 'copy tracks' which is a horrible way to copy animations. Or I did the same thing as you with `ManageAnimations`
Anyway, I am a newbie so I do not know what is the recommended/ideal way to copypaste animations.

**Comment by TheYellowArchitect** [comment]
@Sonix fixed the only code that didn't work, this tutorial is fully functional and can be merged as is.

The only thing that needs improvement is the animation section at the end. Aside of explaining how to copypaste animations properly, the important part is fixing the keys themselves, because currently the player or the mob having full idle/float/moving animations break the game, but anyway, just turning off the animations for player and mobs by not ticking/enabling autoplay on load (button) fixes this.

**Comment by TheYellowArchitect** [comment]
@mhilbrunner Thank you for the reply, you have convinced me that it is indeed the best to force `.webp`
The following argument especially is convincing:

>As is, nobody prevents anyone from opening PRs containing .jpgs or .pngs, those will just hang around until someone takes the time to convert the images; either a maintainer or another contributor salvaging those.

=================

>Also, it's not that important to store lossless originals in case we again switch format years from now and need to re-encode, as most images are screenshots which tend to become outdated and need to be updated before that anyway.

You are right. After `.avif`, it will take many many years, so this problem is basically non-existent after `.avif`

>I haven't spent the time to crunch those numbers, but I definitely feel like a large majority of image contributions are made by recurring contributors, not new/beginner ones.

I am almost certain as well, but you provided strong arguments, so no need to search for accurate numbers

>I agree that in theory, having a system to store lossless binary data somewhere (likely outside this repository), and then compressing the currently best used format (like .webp) automatically could be nice. We don't have that, and we don't have the resources to spend on it.

I didn't mean automating anything for storing current images. Just making a new branch of this repo, and freezing it, so whenever a full conversion is required (`.avif` or `.webp`), someone can make a bash script or something similar to convert images of the same name. Basically, just backing up the current images `.png` and `.jpg`, even though these images probably won't ever be used again (as that would require someone to make a script to automate the process)
tl;dr of the above: Make a backup/duplicate branch before everything is mass-converted to `.webp` just to be safe, and in case these images can be used in the future for `.avif`

**Comment by YuriSizov** [comment]
Seems like this needs a rebase.

**Comment by TheYellowArchitect** [comment]
@YuriSizov I haven't done a rebase before.
I followed the guide here https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

I did pretty much the same, where I did
`git rebase -i a53e8f1326ec84b616ec298671c14963106ebdd2`
Where the commit ID `a53e8f1326ec84b616ec298671c14963106ebdd2` was my first commit (or second? not sure)

Then with vim popup, I saved and closed this:
```
pick 81cec975f Fully finished. COMPLETE.
fixup fd91f8c7d Typo - occured to ocurred
fixup c974b5efa Typo2 - occured occurred
fixup 6caf60b91 Typo3 - occured occurred
fixup 08cf56bd4 Updated camera viewport explanation
fixup 1107cd30c Small polishing1
fixup 6ddcc9ba6 Minor Polishing2
fixup 0607d3e89 Minor Polishing3
fixup 24073e2db Minor Polishing3
fixup 5a0c96721 Minor polishing5
fixup 8272453ed Added missing image reference
fixup 6e2e04cad Minor Polishing 6
fixup f19433921 Minor Polishing 7
fixup 42b30d7d4 Minor Polishing 8
fixup cdff5e2bb Minor Polishing 9
fixup 60204c6ed Minor Polishing 10
fixup fbe9a9a0e small changes to "first 3d game" tutorial GDScripts
fixup 40bc43607 animation panel doesnt require to select node

# Rebase a53e8f132..40bc43607 onto a53e8f132 (18 commands)
#
# Commands:
# p, pick <commit> = use commit
# r, reword <commit> = use commit, but edit the commit message
# e, edit <commit> = use commit, but stop for amending
# s, squash <commit> = use commit, but meld into previous commit
# f, fixup [-C | -c] <commit> = like "squash" but keep only the previous
#                    commit's log message, unless -C is used, in which case
#                    keep only this commit's message; -c is same as -C but
#                    opens the editor
# x, exec <command> = run command (the rest of the line) using shell
# b, break = stop here (continue rebase later with 'git rebase --continue')
# d, drop <commit> = remove commit
# l, label <label> = label current HEAD with a name
# t, reset <label> = reset HEAD to a label
# m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
#         create a merge commit using the original merge commit's
#         message (or the oneline, if no original merge commit was
#         specified); use -c <commit> to reword the commit message
# u, update-ref <ref> = track a placeholder for the <ref> to be updated
#                       to this position in the new commits. The <ref> is
#                       updated at the end of the rebase
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
```

Then did git push --force pretty much as in the guide. Did I do the rebase good and without mistakes?

**Comment by YuriSizov** [comment]
> Where the commit ID `a53e8f1326ec84b616ec298671c14963106ebdd2` was my first commit (or second? not sure)

You're supposed to rebase against the head (the latest commit) of the master branch, not against your first commit. As you can see at the bottom of this page, you still have merge conflicts. That's because your branch and the master branch have diverted too much for automatic conflict resolution to work if we were to merge it. This is why you need to rebase against the head of the master.

You might need to fix conflicts as you go, because your changes may overlap with the changes done in master since you branched out. When you are presented with that choice, you can either pick the version of the code from the master branch, or your own version, or merge the two together in some way, or replace the conflicting bit with some entirely new text. It's up to you. You may need to do this for every commit in your branch, and after each commit you would need to stage your changes and run `git rebase --continue` (or an equivalent action in a GUI app if you use one).

**Comment by TheYellowArchitect** [comment]
Thank you for the feedback, it helps me learn how to rebase and succeed in rebasing. But I need a bit more help so as to not fail again. Even after `git fetch` on either of the branches `TheYellowArchitect:squash` or `TheYellowArchitect:master`, my `git log` returns all entries up to september 29 (which I guess is when my last commit here was made), so I cannot rebase against the latest `godotengine:master` commit. I am not very good on git, so I don't know how to fix this, since I think I have totally diverted from the upstream branch.

That said, for diff editing, I don't use any kind of GUI like VS Code, so I have to ask: If I resolve the conflicts via the 'Resolve Conflicts' button here in Github, won't that work and avoid the need of doing a rebase?

**Comment by Sonix** [comment]
@TheYellowArchitect If you don't mind I could try a rebase of your branch and link it here.

**Comment by TheYellowArchitect** [comment]
@Sonix I thank you for your offer, and I certainly cannot resist it, since I really want this to be merged so as new users won't have compile errors or try to fix what is already fixed here, and especially considering I may fail the rebase hence making my efforts be in vain

I thank you in advance

**Comment by Sonix** [comment]
@TheYellowArchitect I've rebased my fork of your repository on top of the current `master` branch of `godot-docs`. I can not force push or create a pull request that automatically force-pushes onto your repo, so I'd recommend you do the following, assuming you have a git command line tool available:

First, make sure you're in the right directory and have the `squash` branch checked out.

1. `git remote add sonix git@github.com:Sonix/godot-docs.git` This adds my fork of your repo as a new remote to your git repo (You might need to change the URL to another option from https://github.com/Sonix/godot-docs if you don't use SSH authentication for Github)

2. `git fetch sonix` This pulls all the branches and commits from my fork and stores them under sonix/*

3. `git reset --hard sonix/squash` This *resets* (replaces) your current branch (here: `squash`) with my version of that branch, where I did the rebase
4. `git push origin --force` this force-pushes your current squash branch (which is now the same as my rebased version) to Github, into the `squash` branch.

On Github, your branch should now show all the changes from godot-docs/master as well as the commits from your PR on top if it. This PR should update and indicate it now longer has merge conflicts.

**Comment by TheYellowArchitect** [comment]
Thank you very much, I did all the steps as you said and stumbled into no problems. Even the SSH authentication in step 1 was as you said, and I just did 
```git remote add sonix https://[SSHKEY]@github.com/Sonix/godot-docs.git```

As for the last step, `origin` was undefined, and I replaced it with
```https://[SSHKEY]@github.com/TheYellowArchitect/godot-docs``` and it seems to have worked, thanks for the help! <3

**Comment by Sonix** [comment]
Good to hear! Lets hope it finds it way into the main repo now :)

**Comment by YuriSizov** [comment]
> Even after `git fetch` on either of the branches `TheYellowArchitect:squash` or `TheYellowArchitect:master`, my `git log` returns all entries up to september 29 (which I guess is when my last commit here was made), so I cannot rebase against the latest `godotengine:master` commit.

So, you have a fork of this repository, where you create and manage branches — `TheYellowArchitect/godot-docs`. This fork has no connection to the original repository, which is `godotengine/godot-docs`. When changes in our repository happen yours doesn't receive them automatically, because these are two different repositories.

When you work locally, these repositories would be called "remotes", as in a remote server, a remote repository. By convention, your own fork, `TheYellowArchitect/godot-docs`, is called `origin`, and it's the default name often suggested by git applications. Our original repository, `godotengine/godot-docs`, is called `upstream`. We recommend sticking to this convention, as it is used throughout the documentation and discussion and is typical for other open source projects as well.

So you should have 2 remotes on your local machine. One called `origin` which is your default remote and should point to your own repository. And one called `upstream` which should point to this repository. In order to get the latest changes you need to fetch from `upstream`, or just from all your remotes together. Then you will be able to see commits from all remotes, and from your local branches, and can rebase your local branch against the latest commit in the upstream master branch.

Your own master branch can be updated in a similar way, but instead of updating you can just pull into your local master branch and push it to the `origin` remote.

-----

If it sounds complicated, don't worry. With the help from Sonix you've already started working with multiple remotes. Just instead of using their repo you can use the `upstream` which is `godotengine/godot-docs`, and `origin` which is `TheYellowArchitect/godot-docs`.

**Comment by TheYellowArchitect** [comment]
Thank you for the elaborate explanation, I am confident I can pull off a rebase in the future using your 2 replies as reference. I had no idea that I can git fetch another repo (which I had branched off)

**Comment by mhilbrunner** [comment]
cc @NathanLovato Any input on this? If you're too busy, thats fine of course.

Otherwise, I'd like to get this merged before our upcoming docs sprint to avoid too many conflicts/rebases, even if further iteration/changes happen afterwards, that can be done in further PRs.

File: getting_started/first_3d_game/06.jump_and_squash.rst [hunk 2]
```diff
@@ -275,30 +280,30 @@ With this code, if no collisions occurred on a given frame, the loop won't run.
...
+``get_collider`` property to check if we collided with a "mob" by calling
+``is_in_group()`` on it: ``collision.get_collider(index).is_in_group("mob")``.
 
 .. note::
 
     The method ``is_in_group()`` is available on every :ref:`Node<class_Node>`.
 
 To check that we are landing on the monster, we use the vector dot product:
-``Vector3.UP.dot(collision.normal) > 0.1``. The collision normal is a 3D vector
+``Vector3.UP.dot(collision.get_normal) > 0.1``. The collision normal is a 3D vector
```
Comment (akien-mga) [inline getting_started/first_3d_game/06.jump_and_squash.rst:298]: Does this work?

File: getting_started/first_3d_game/05.spawning_mobs.rst [hunk 3]
```diff
@@ -234,20 +238,21 @@ Let's code the mob spawning logic. We're going to:
 .. tabs::
  .. code-tab:: gdscript GDScript
 
-   func _on_MobTimer_timeout():
-       # Create a new instance of the Mob scene.
-       var mob = mob_scene.instantiate()
+   func _on_mob_timer_timeout():
+        # Create a new instance of the Mob scene.
+        var mob = mob_scene.instantiate()
```
Comment (akien-mga) [inline getting_started/first_3d_game/05.spawning_mobs.rst:243]: Indentation inconsistency (3, then 8).

Looks like most code blocks here were wrongly indented with 3 spaces, leading to inconsistencies. This should be reviewed.

File: getting_started/first_3d_game/06.jump_and_squash.rst (hunk 2 above)
Comment (TheYellowArchitect) [reply getting_started/first_3d_game/06.jump_and_squash.rst:298]: yup, just confirmed it again in Godot beta10

File: getting_started/first_3d_game/06.jump_and_squash.rst [hunk 4]
```diff
@@ -136,7 +134,7 @@ called ``move_and_slide()``.
 
        # Jumping.
        if is_on_floor() and Input.is_action_just_pressed("jump"):
-           velocity.y += jump_impulse
+            target_velocity.y = jump_impulse
```
[the remark this answers is not in the document]
Comment (TheYellowArchitect) [reply getting_started/first_3d_game/06.jump_and_squash.rst:137]: Nice eye

**Comment by TheYellowArchitect** [comment]
@akien-mga I fixed the reported indentation issues

File: getting_started/first_3d_game/06.jump_and_squash.rst (hunk 2 above)
Comment (YuriSizov) [reply getting_started/first_3d_game/06.jump_and_squash.rst:298]: Nope, this is still a typo. It should be `Vector3.UP.dot(collision.get_normal())`. The code in the examples is correct, but here it is wrong. You probably tested the examples, which don't have a typo.

Comment (TheYellowArchitect) [reply getting_started/first_3d_game/06.jump_and_squash.rst]: You are right, I missed the `()` and as you said, I tested the example and saw `get_normal` working so I wrote the above lol

Anyway, fixed in latest push :+1:

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
There's some remaining issues and things that could be changed (i.e. there's still some whitespace weirdness in code examples, mostly from before this PR), but everything looks pretty good to me besides that.

Great work! I'm going to go ahead and merge this now, we can do further work in smaller PRs. Otherwise, this is going to end in rebase hell with the docs sprint now started. :) Thanks for this large rework!

**Comment by TheYellowArchitect** [comment]
It's the least I can do, since I am not skilled enough to help with engine bugfixes, yet want to help other gamedevs. Also this pull request would be incomplete without @Sonix <3

**Comment by NathanLovato** [comment]
Sorry @mhilbrunner, you pinged me right when moving house, I was AFK for the past couple of days. Thanks for the work @TheYellowArchitect ! And congrats on getting this merged.

File: getting_started/first_3d_game/04.mob_scene.rst [hunk 5]
```diff
@@ -172,24 +173,21 @@ between ``-PI / 4`` radians and ``PI / 4`` radians.
...
 .. tabs::
  .. code-tab:: gdscript GDScript
 
    func initialize(start_position, player_position):
        # ...
 
-       # We calculate a random speed.
-       var random_speed = rand_range(min_speed, max_speed)
+       # We calculate a random speed (integer)
+       var random_speed = randi_range(min_speed, max_speed)
```
Comment (aaronfranke) [inline getting_started/first_3d_game/04.mob_scene.rst:186]: This is wrong, this should be a float.

1. This value is always used in a float context.
2. It's perfectly sensible for a mob to have a speed of `10.5` meters per second for example.
3. This was float before, `rand_range` in Godot 3.x uses floats.

Comment (TheYellowArchitect) [reply getting_started/first_3d_game/04.mob_scene.rst:186]: I cannot re-open this now that it's merged, I suggest opening a new issue or doing a pull request, its pretty small anyway
